internal/testtypes/matchers.go

```diff
@@ -0,0 +1,274 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testtypes
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/onsi/gomega/format"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// This code is a modified version of the mergePatch code from the controllers/topology/internal/mergepatch pkg.
+// TODO: Create a shared patchdiff library that can be shared for both patching and matching.
```

Comment:
Replicating my comment from the old PR:
> Are we sure we want a separate package for that? Sometimes a little bit copying is better then a new package/dependency

Reply:
Replicating my response 🙃  (sorry I didn't notice the bad branch name earlier!) 

> Copying is definitely the right way for now - but having this code in multiple places doesn't seem like the right solution in the long term as its usage might grow - especially if mergepatch goes out of internal. I think having an issue for this is a good idea, and we can decide down the line if we want to actually execute on it.

Reply:
Sorry looked at the wrong place for the response and didn't see it :/ 

Might be a good idea to open an issue for it. Not sure how the usual "process" is. TODO in the code or issue or both

Reply:
I think maybe we leave the TODO out of the code and open an issue - I always feel like a TODO should focus on something more short term, whereas this might not be resolved for a while. An issue (at least with the Kubernetes bots keeping us honest) forces us to make a decision on how to proceed on a regular basis.